### ISSUE TITLE
bug: CRITICAL statement in IF statement triggers "Unexpected end of scope at line"

### ISSUE THREAD

**Issue by han190** [fortran-lang/fortls#255]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Description

As described in the title, when `CRITICAL` statement appears in `IF` statement, an error "Unexpected end of scope at line" is triggered. 

### Screenshots

![Screenshot 2023-01-11 193437](https://user-images.githubusercontent.com/13799118/211948421-fabbbf8f-8a6d-488b-ab93-7b0e1d68beee.png)

### Expected Behaviour

I believe the test code should be standard conforming.

### Version of Modern Fortran

v3.4.2023011101

### Version of Visual Studio Code

v1.74.3

### Platform and Architecture

WSL

**Comment by gnikit** [comment]
See PR https://github.com/fortran-lang/fortls/pull/256 (will merge once green).

I will wait a bit before I release a new version since the current `master` of fortls contains a bunch of new features that I need to go over again to make sure I don't disrupt people's workflows.

You can test the new version and tell me if you see anything wrong, including this bug by using:

```sh
pip install git+https://github.com/fortran-lang/fortls.git
```

**Comment by han190** [comment]
Thank you, the issue is resolved so fast! In #256 I saw that you combined `block` and `critical` in the same regex pattern. I guess this is fine for a quick fix, but based on the standard a BLOCK construct is
```
R1107 block-construct    is block-stmt
                         [ block-specification-part ]
                         block
                         end-block-stmt
R1108 block-stmt         is [ block-construct-name : ] BLOCK
```
and a CRITICAL construct is
```
R1116 critical-construct is critical-stmt
                            lock
                            end-critical-stmt
R1117 critical-stmt      is [ critical-construct-name : ] CRITICAL [ ( [ sync-stat-list ] ) ] 
```
So syntactically speaking a CRITICAL construct may include an extra `[ ( [ sync-stat-list ] ) ] `. Like what we discussed in #172, many of the coarray issues encountered require a better Fortran parser. And, yes I will keep you posted if I spotted new issues.
